File: services/auth-service/src/schemas/organization.py
from pydantic import BaseModel, ConfigDict, Field
from uuid import UUID
from typing import Optional
from datetime import datetime
from ..models.enums import OrganizationPrivacy, OrganizationStatus, OrganizationJoinPolicy
from ..schemas.user import UserOrgMember
from ..models.enums import MemberStatus

class OrganizationBase(BaseModel):
    name: str = Field(..., min_length=3, max_length=255)
    description: Optional[str] = None
    logo_url: Optional[str] = None
    privacy: OrganizationPrivacy = OrganizationPrivacy.PUBLIC

    model_config = ConfigDict(from_attributes=True)

class OrganizationCreate(OrganizationBase):
    pass

class OrganizationUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    logo_url: Optional[str] = None
    privacy: Optional[OrganizationPrivacy] = None
    join_policy: Optional[OrganizationJoinPolicy] = None

    model_config = ConfigDict(extra='ignore')

class OrganizationGetPublic(OrganizationBase):
    id: UUID
    slug: str
    owner_id: UUID

class OrganizationResponse(OrganizationGetPublic):
    status: OrganizationStatus
    join_policy: Optional[OrganizationJoinPolicy] = None
    created_at: datetime
    updated_at: datetime

class OrganizationWithRole(OrganizationGetPublic):
    role: str

    model_config = ConfigDict(from_attributes=True)

class OrganizationAdminWithRole(OrganizationResponse):
    role: str

    model_config = ConfigDict(from_attributes=True)


class UpdateJoinPolicyRequest(BaseModel):
    join_policy: OrganizationJoinPolicy


class OrganizationMemberResponse(BaseModel):
    id: UUID
    user: UserOrgMember
    status: MemberStatus
    joined_at: datetime
    is_owner: bool = False

    model_config = ConfigDict(from_attributes=True)


class MembersListResponse(BaseModel):
    total: int
    members: list[OrganizationMemberResponse]


class OrganizerResponse(BaseModel):
    id: UUID
    user: UserOrgMember
    added_at: datetime

    model_config = ConfigDict(from_attributes=True)


class OrganizersListResponse(BaseModel):
    total: int
    organizers: list[OrganizerResponse]


class TeamOverviewResponse(BaseModel):
    owner: UserOrgMember
    organizers: list[OrganizerResponse]
    members: list[OrganizationMemberResponse]
    total_members: int
    total_organizers: int


class TransferOwnershipRequest(BaseModel):
    new_owner_id: UUID

File: services/frontend/types/organization.ts
export enum MemberStatus {
  PENDING = "PENDING",
  ACTIVE = "ACTIVE",
  INVITED = "INVITED",
  BANNED = "BANNED",
}

export enum OrganizationStatus {
  PENDING = "PENDING",
  ACTIVE = "ACTIVE",
  REJECTED = "REJECTED",
  SUSPENDED = "SUSPENDED",
  EXCLUDED = "EXCLUDED",
}

export enum OrganizationPrivacy {
  PUBLIC = "PUBLIC",
  PRIVATE = "PRIVATE",
}

export enum OrganizationJoinPolicy {
  INVITE_ONLY = "INVITE_ONLY",
  REQUEST_ONLY = "REQUEST_ONLY",
  LINK_ONLY = "LINK_ONLY",
  INVITE_AND_REQUEST = "INVITE_AND_REQUEST",
  INVITE_AND_LINK = "INVITE_AND_LINK",
  REQUEST_AND_LINK = "REQUEST_AND_LINK",
  ALL = "ALL",
}

export enum OrgRole {
  OWNER = "OWNER",
  ORGANIZER = "ORGANIZER",
  MEMBER = "MEMBER",
}

export interface UserOrgMember {
  id: string;
  username: string;
  email: string;
  first_name: string | null;
  last_name: string | null;
  avatar_url: string | null;
}

export interface OrganizationBase {
  name: string;
  description: string | null;
  logo_url: string | null;
  privacy: OrganizationPrivacy;
}

export interface OrganizationCreate {
  name: string;
  description?: string;
  privacy: OrganizationPrivacy;
  logo?: File;
}

export interface OrganizationUpdate {
  name?: string;
  description?: string;
  logo_url?: string;
  privacy?: OrganizationPrivacy;
  join_policy?: OrganizationJoinPolicy;
}

export interface OrganizationGetPublic extends OrganizationBase {
  id: string;
  slug: string;
  owner_id: string;
  created_at: string;
}

export interface OrganizationResponse extends OrganizationGetPublic {
  status: OrganizationStatus;
  join_policy: OrganizationJoinPolicy | null;
  created_at: string;
  updated_at: string;
}

export interface OrganizationWithRole extends OrganizationGetPublic {
  role: string;
}

export interface OrganizationAdminWithRole extends OrganizationResponse {
  role: string;
}

export interface OrganizationMemberResponse {
  id: string;
  user: UserOrgMember;
  status: MemberStatus;
  joined_at: string;
  is_owner: boolean;
}

export interface MembersListResponse {
  total: number;
  members: OrganizationMemberResponse[];
}

export interface OrganizerResponse {
  id: string;
  user: UserOrgMember;
  added_at: string;
}

export type OrganizationOrganizerResponse = OrganizerResponse;

export interface OrganizersListResponse {
  total: number;
  organizers: OrganizerResponse[];
}

export interface PendingMemberRequest {
  id: string;
  user: UserOrgMember;
  status: MemberStatus;
  created_at: string;
  updated_at: string;
}

export interface PendingRequestsResponse {
  total: number;
  requests: PendingMemberRequest[];
}

export interface TeamOverviewResponse {
  owner: UserOrgMember;
  organizers: OrganizerResponse[];
  members: OrganizationMemberResponse[];
  total_members: number;
  total_organizers: number;
  created_at: string;
}

export interface UpdateJoinPolicyRequest {
  join_policy: OrganizationJoinPolicy;
}

export interface TransferOwnershipRequest {
  new_owner_id: string;
}

export type OrganizationListItem = OrganizationWithRole | OrganizationAdminWithRole;

export interface OrganizationFormData {
  name: string;
  description: string;
  privacy: OrganizationPrivacy;
  logo?: File | null;
}

export interface OrganizationSettingsFormData {
  name: string;
  description: string;
  privacy: OrganizationPrivacy;
  join_policy: OrganizationJoinPolicy;
  logo?: File | null;
}
